### DISCUSSION TITLE
Infer Immer middleware type for `setState` from `setState` instead of `getState`

### DISCUSSION THREAD

**Discussion by chrisvander** [pmndrs/zustand#1621] [General] [answered]
The Immer middleware infers the type nextStateOrUpdater for setState from getState:

```ts
type StoreImmer<S> = S extends {
    getState: () => infer T;
    setState: infer SetState;
} ? SetState extends (...a: infer A) => infer Sr ? {
    setState(nextStateOrUpdater: T | Partial<T> | ((state: Draft<T>) => void), shouldReplace?: boolean | undefined, ...a: SkipTwo<A>): Sr;
} : never : never;
```

This is typically not a problem unless you're using a middleware like [chrisvander/zustand-computed](https://github.com/chrisvander/zustand-computed), where the state returned by getState differs from what is allowed to be set via setState. In the case of the computed middleware, getState returns additional keys beyond what the user is allowed to modify. Can we change the Immer state to infer off of the SetState function? E.g. use something like `Parameters<SetState>[0]`.

**Comment by dai-shi** [comment]
`Parameters<SetState>[0]` sounds fairly reasonable, but then we can't support devtools action name with immer middleware, can we?

**Reply by chrisvander** [reply]
Sorry, split the discussion by opening the issue at the same time. Not sure what you mean, how would this type change affect devtools?

**Reply by dai-shi** [reply]
Here's a repro: https://tsplay.dev/WKpqZw

**Reply by chrisvander** [reply]
We could use the `A[0]` type as well. Thanks for the reproduction. That might be something specific to my library, from a functionality perspective I can't see why it wouldn't work. Putting `immer` outside `zustand-computed` does seem to allow the action name to pass through.

**Reply by dai-shi** [reply]
`A[0]` still doesn't work, right? Anyways, I hope you will find a solution.

**Reply by chrisvander** [reply]
Interestingly, it does work fine when you explicitly define the Middleware before/after deps. Immer seems to be losing them when held inside of `zustand-computed`. Is it the way I'm extending the middleware declarations?

[Playground Link](https://www.typescriptlang.org/play?noImplicitAny=false&ts=4.9.5#code/JYWwDg9gTgLgBAbzgYygUwIYzXAvnAMyghDgCIAvAVwGcYMA7AEzIChRJZE4m0A3GBAgAbGnkLFSlWvWYB6EMCZNhaAO4Z0bDtHhJQINFHFES5anUZMFSles1o5Bo9vC7uyEmCrYmJyeYyVgC0nuA+aCysrDAAnmA4AMqC6HAAvIiscCgQVAwwAFxwDFQgAEZGWXDADMhFABQAlOkAfHB8EEpVvHVwTa3tnUysuNFxCXAAwl4RTMnQOBkIVZ55MIkAjkUl5ZWjrAR5yDDAEAw54djJWGj1lthF8+iNRdOXkU84y9noMFRQ52+2Rya02RXuaAAdKt8nAAFRwuAAJgANFVRvtPAw6HBaGhPukUOgbgAeT4tJr1br8QQiGhU4EXby+BmMvo0NAwZppNr1IFskH5IoARjRAuyNV6-R5cA5MHqd3o2G5vKQMMKsqVUPVcAA1HBhXhGijCBhRGgTQBySWW41VAU9Boq2WchUQ518wUaiHQ3Kw4IGo0mghmjlWnq2sVs3B2gVhZn4rX2uCNKqp1OsLE4vGfACSIEMxgyqEw2DJKTQFMarN4AiEolZ2Xjs0bwOcUHLCxNAG1u9JLPJa7TRGQTQx+EYALo9sjIAAWUGAND4ViMcgssiYoRmvlHUx3Hwrk+ncF7-c3TgLLjHE6gx4pd05Hv5bPVIqjAslTtayYFcrdWrOi+4pwHIchwGgAAeGDgKocBqMAMBznA+aFnAijKKoGjoL+4o+jquoZMKuGMrgwahhacDWrUkYkTwaBSs6-6Kjcz5euCWq+mscABoaMbkea4YMbRAoxrGr4Htc2DJqm2TptEWbwDmFaoUYADqiFzrkMAAKKQWAwjAMgiGEiWpLkpS1J1nSrbNiyuFgdUDAEEYYhITg4w4J26A9ie3aTi0uHtm6T4DHydFvgaH5sl+fQqnR2TMe6AzAeKjlQTBBk4AhSEoVexgYXY2FoAlwL4X68CEQapVkaaglUTa4kOgx34yklgFhWqFUcTcXH+oG-F1WGVERk1wJiSauF2YmNwyWmrCpkAA)

**Reply by chrisvander** [reply]
The middleware declaration uses a second generic parameter, `A`. Does that mess anything up?

```ts
type StoreCompute<S, A> = S extends {
  getState: () => infer T
  setState: infer SetState
}
  ? Write<StoreApi<T & A>, { setState: SetState; getState: () => T & A }>
  : never

type WithCompute<S, A> = Write<S, StoreCompute<S, A>>

declare module "zustand" {
  interface StoreMutators<S, A> {
    "chrisvander/zustand-computed": WithCompute<S, A>
  }
}
```

**Reply by dai-shi** [reply]
https://tsplay.dev/mL3OaN
It doesn't seem only an issue with immer middleware. All other middleware don't seem to work.
Can you try something simple and see if it works as expected?

**Reply by chrisvander** [reply]
So the `setState` being inferred from Immer is still a problem, but you're right that the generics aren't being passed down appropriately. 

[Try this playground.](https://www.typescriptlang.org/play/?noImplicitAny=false&ts=5.5.4#code/JYWwDg9gTgLgBAbzgYygUwIYzQGjgZRizQGF0to9Do0BBMYKmGgWQFcjmoBJAEzQB2MYADNgaKHAC+cEVAgg4AIgBebAM5EBvJQChQkWIjj8AbswgAbdXnVsARutTB7aAOrAYAC3xpLaZC48ZAV7YAE0aVl5RVUNLV4AehBgXl5-AHcMdD0DaHgkUBAJKLkFZTVNDG1k1PS0LPREook9XRgATzBI6nQ4AF5EXTgUCDYhAC44ATYQVyhhuHDkKYAKAEoBgD44UwhUxf4VuA3t3f3eXSlddq7IkgUwDjReXsjBhEWQ8Zh8AEcpjM5hIrjcRONAsAIAJRuBnoRiKsqtgpm91lMHnDsK8uJFPiN0DA2FAYfiRqMfv8psi0AA6b5COAAKiZcAATDhFtdrro0AAPQzwTrdOCYp7YhHYADyYBg6gAPAAVHYfRYAazQHXUAH41hqOhARHBFesANoAXUOwHUGHs-gACvI+R1dXB7BArJgBIs0H82BhLJ4OgAxASu+UATS2qwwUwjeHscc2-R27s91VBvIF+Tgwvuj2eOOIZEwXAGcHli0VcH52G06jgEHsACsAjBOSNaDW+XXeA2m63Ah24CwwA3a4I+3BTW92JxoHxBMIxBI8OM1QIIBkBOaLeWLcOWMhxz3Jw2Z7i5xQePwhKJxFA1wIN1ud3vBgfdNHFuFPMAA8AKgSKinCkOQXBKngpq0jBo42NOSjIF4UDWqY1T8FAiSVAkAC0IRYi8Sh4LQ5rmngR7qFsw74eKaBrDSUwmmctDDhAso6hiBYSqBMpykqVG6MmOySmBpbQJBI5jlBpqIchqHoRIWHxOheFcYRxFkXAMG0hRmnVgAZHAtBbDceZwG4KHYBJACqKpwFKKQwBJ+qGnAtlwIZ1m3CKbxis88r4MRdn4N2vYNmSADmaC-KBaxCUsAgiCUiqgiM2r2Y5AW4vQwBKh5RlUco6jRSJSgmSMgJoKYIJmR43h+VZgUFeWFmeGgAVMDQDXtU1xkmd5kRdoMdVeN1WU0Dl43oIV3U4jQJlmSFgy0DJxUxcQSiWroRyWNkkQgBAvBsP4FTKTUaECMAli7UoQwjOE2BQCIGDID0l4cNeCq9TsZIjLJKHqBdGFKVU2iqQROhTCNY3fVymb8oKowCJosK0bw5YAIxwBgDbrpu27Yw2s0iSW143CEyPwBoaBvOWqClj1uLRusqyHFVFjWKz5Ko4WXPc0sIDFFAqxItF8WrL9-MMjAUwY8O-MJccpwpnAa0izS4tINL1KgfSYyMgA1HAWNSOseDPdYuBwAA5Ms1tm4sCtHHFZxq0ioGaxSkyq7r0twDhxvSGbsgBsVeDW0c9vy+SpsOwrNHwqBjtwOsiyp6nugUyj1NvNwgslIM9PEFNaDM3zZgc+ofMjAn2LV+SLRQCX0lxKDSQVx61hEdMVUSJpMlIQDQOKdhKm1+popqXN6CkS3o81I33cRNVUCkdGoswJryfc9rxvR9zywuym2-827GtnJLCsjIkiTdhg4AnRknheHAedC3AKRpJke0nwrNJ6z8OABtBgY1-jHc2ocra2wEMgKOYDnYnHimfD2ZwJZexlj7YgADGQBxNsHC2YcbaRzjvzWO+8ebYBEsnVOIx07k2hNnYqud85QBGmMGAABRAUgZkCeDpuBRm80Njl3Zp3Kuydx68HrvdFh6sxaoMvnAG+cAAB62owG7zlmAw+iDthgJGMg4g8VFEK2Ufye+YBH7P1fiwj+dRv7oH0eSf+ftgHGycVICBltw52xIU7AIR8diGOwJ7XeLj9bwFwUHLxhCI4BDgQrMhcAT7jyodzGhKdBI3EzgwqmTDcRvwkGyfhDMS5lzZuYMR9dJHSIFkLZu04ZLz3bqIqw6gl691XlBf68ltAjzOrwcGaNu7E1xLPac5p15rS3qYxIaiNERNluQ+6MDAlOOCWgYxTjr633MQ-SIT9vA2Pfp-eojQ0DbMwdgbB8A3GgKvqfX2ET-hAJAR4mJUDfHLJMAE3RKsNmhMWVcukfsommw+eHYh5CkkpLUmk8kGS6E5MpnAHOBSWEAGYSnFzeOUkYHc2nVKnrUuwjhnCuBGr4fwgRoByM3gok+yj5kPM0d8nRytyoPNVtFdWKCVYmP5oSYkpJ0E6ywa4wO1wuXSAhTbL58Dfkcu5TAXlRiFGiuBTc-2kr8GQMhfEvxMd1jB1hQReFtC05ZORYwmm6KhYABZsWNSZsIiplciUQxJQ4JwKEKXPypW2WlG8Zn8yZeollQKtEPPZUJdZPL3Zqv5ZcoVJJjBhKeYAo2Jt3kh28XKmBCSr4IKVQC9V6bxURO1Xg2VcTYGGpGDC+OcKk7pMtRnLOeTbU0EKVAAArE6wR01XX4taZzCRxLTVhAiGgzRMqTjTIZbM1R4ar4xr0dK0tSbpUEmisKtNQLwmZslTmghnyC31vJMWpB8bz4qxnQejNOCdU1qhSfRtUtm3EGoW2oAA)

On line 116 you'll notice that `set` contains the `getState` type with the computed store, which I'm trying to avoid. Regardless, I manually typed the Immer middleware for `useStoreImmer` and `useStoreImmer2` and that mitigated the problem you highlighted above. It seems that the `Mos` type is not being passed down.

**Reply by dai-shi** [reply, marked as answer]
https://tsplay.dev/WGgq9W
Now, I know why. `ComputedStore` has to be inferred before passing to the next middleware.

**Reply by chrisvander** [reply]
Beautiful. Thanks. I suppose I could use currying, either way I see now that the computed state needs to be inferred first.
